### PR TITLE
[v6r17] Fixes in Stager and DataManager

### PR DESCRIPTION
**Pull request by phicharp**
* Stager: mostly streamline code, but also monitor files even when there is no requestID (dCache returns none when staging a file that is already staged!)

* DataManager: add key argument forJobs (default False) in getReplicas() in order to get only replicas that can be used for jobs (as defined in  the CS)
Use this feature in the API as well (default unchanged except for splitting where it is set to True)

### CODE REVIEW

File: StorageManagementSystem/Agent/StageMonitorAgent.py [hunk 1]
```diff
@@ -62,40 +60,43 @@ def __monitorStorageElementStageRequests( self, storageElement, seReplicaIDs, re
 
     # Since we are in a given SE, the LFN is a unique key
     lfnRepIDs = {}
-    lfnReqIDs = {}
     for replicaID in seReplicaIDs:
       lfn = replicaIDs[replicaID]['LFN']
       lfnRepIDs[lfn] = replicaID
-      requestID = replicaIDs[replicaID].get( 'RequestID', None )
```
Comment (phicharp) [inline StorageManagementSystem/Agent/StageMonitorAgent.py]: This is where the main change is: no test any longer on requestID

File: StorageManagementSystem/Agent/StageRequestAgent.py [hunk 2]
```diff
@@ -222,23 +216,23 @@ def _getOfflineReplicas( self ):
...
+    # Set it if not yet done
+    self.storageElementUsage.setdefault( storageElement, {'TotalSize': 0.} )
     return self.storageElementUsage[storageElement]['TotalSize']
 
   def __cache( self, storageElement ):
     """ Retrieve cache size for SE
     """
-    if not storageElement in self.storageElementCache:
-      self.storageElementCache[storageElement] = gConfig.getValue( "/Resources/StorageElements/%s/DiskCacheTB" % storageElement, 1. ) * 1000. / THROTTLING_STEPS
+    self.storageElementCache.setdefault( storageElement,
```
Comment (chaen) [inline StorageManagementSystem/Agent/StageRequestAgent.py]: remember that when calling setdfefault, the second argument is always evaluated, even if not needed. So it is maybe an overkill here ?

File: StorageManagementSystem/scripts/dirac-stager-monitor-requests.py [hunk 3]
```diff
@@ -22,160 +22,157 @@
...
+
+  for key in ( 'status', 'se', 'limit' ):
 
     if not key in switches:
       print "You're not using switch --%s, query may take long!" % key
-      
-  if 'status' in  switches.keys():  
-    if not switches[ 'status' ] in ( 'New', 'Offline', 'Waiting','Failed','StageSubmitted','Staged' ):
+
+  if 'status' in  switches.keys():
```
Comment (chaen) [inline StorageManagementSystem/scripts/dirac-stager-monitor-requests.py]: ```if 'status' in switches:```

File: StorageManagementSystem/scripts/dirac-stager-monitor-requests.py [hunk 4]
```diff
@@ -22,160 +22,157 @@
...
+    if not switches[ 'status' ] in ( 'New', 'Offline', 'Waiting', 'Failed', 'StageSubmitted', 'Staged' ):
```
Comment (chaen) [inline StorageManagementSystem/scripts/dirac-stager-monitor-requests.py]: ```if switches['status'] not in ... ```

File: StorageManagementSystem/scripts/dirac-stager-monitor-requests.py [hunk 5]
```diff
@@ -22,160 +22,157 @@
...
   # weird: if there are no switches (dictionary is empty), then the --limit is ignored!!
   # must FIX that in StorageManagementDB.py!
   # ugly fix:
-  newer = '1903-08-02 06:24:38' # select newer than 
-  if 'limit' in dictKeys:
-    print "Query limited to %s entries" %switchDict['limit']   
-    res = client.getCacheReplicas(queryDict, None, newer, None, None, int(switchDict['limit']))
+  newer = '1903-08-02 06:24:38'  # select newer than
+  if 'limit' in switchDict:
+    print "Query limited to %s entries" % switchDict['limit']
```
Comment (chaen) [inline StorageManagementSystem/scripts/dirac-stager-monitor-requests.py]: gLogger ?

File: DataManagementSystem/Client/DataManager.py [hunk 6]
```diff
@@ -1524,6 +1524,25 @@ def __filterTapeReplicas( self, replicaDict, diskOnly = False ):
...
     return
 
+  def __filterReplicasForJobs( self, replicaDict ):
+    """ Remove the SEs that are not to be used for jobs, and archive SEs if there are others
+    The input argument is modified
+    """
+    seList = set( se for ses in replicaDict['Successful'].itervalues() for se in ses )
+    # Get a cache of SE statuses for long list of replicas
+    seStatus = dict( ( se, ( self.dmsHelper.isSEForJobs( se ), self.dmsHelper.isSEArchive( se ) ) ) for se in seList )
+    for lfn, replicas in replicaDict['Successful'].items():  # Beware, there is a del below
```
Comment (chaen) [inline DataManagementSystem/Client/DataManager.py]: nice comment !! :D

File: StorageManagementSystem/Agent/StageRequestAgent.py (hunk 2 above)
Comment (phicharp) [inline StorageManagementSystem/Agent/StageRequestAgent.py]: Perfectly right ;-)

**Comment by chaen** [comment]
As a general comment, there starts to be, IMHO too many options to the getReplicas call, and I fear that we will only go in the direction of adding some. Could we maybe have a more generic approach ? Because in the end, what we do is to get all the replicas, and then filter them away depending on whether they are on disks only, whether the SE is active, whether it is for a job, etc. I am sure that there is something to be done here with the filter function... :-) 
But to come back on this specific 'forJob', is the DataManager really a place to care about what the data is used for ? This disturbs me a bit from the logic point of view

File: StorageManagementSystem/scripts/dirac-stager-monitor-requests.py (hunk 3 above)
Comment (phicharp) [inline StorageManagementSystem/scripts/dirac-stager-monitor-requests.py]: Done

File: StorageManagementSystem/scripts/dirac-stager-monitor-requests.py (hunk 4 above)
Comment (phicharp) [inline StorageManagementSystem/scripts/dirac-stager-monitor-requests.py]: Done

File: StorageManagementSystem/scripts/dirac-stager-monitor-requests.py (hunk 5 above)
Comment (phicharp) [inline StorageManagementSystem/scripts/dirac-stager-monitor-requests.py]: Done for all print

**Comment by phicharp** [comment]
A matter of taste, but anyway it has to be done somewhere... The core of LHCbDirac is never using the API, therefore having it in there is not enough. Rather than replicating the code, why not in the DataManager which is already an aggregate of FC and StorageElement? It doesn't harm as long as the default is to get everything...
We can have getReplicasForJobs() that would call getReplicas() and then apply the filter, but is it a gain?
I have no strong opinion... The problem will start if we want to have all cross-products of flags...

File: Interfaces/API/Dirac.py [hunk 7]
```diff
@@ -1022,7 +1022,7 @@ def __printOutput( self, fd = None, message = '' ):
   #       print self.pPrint.pformat( metaDict )
 
   #############################################################################
-  def getReplicas( self, lfns, active = True, preferDisk = False, diskOnly = False, printOutput = False ):
+  def getReplicas( self, lfns, active = True, preferDisk = False, diskOnly = False, forJobs = False, printOutput = False ):
```
Comment (fstagni) [inline Interfaces/API/Dirac.py]: as a safe measure I'd put the new parameter as last.

File: DataManagementSystem/Client/DataManager.py [hunk 8]
```diff
@@ -1616,11 +1637,27 @@ def getReplicas( self, lfns, allStatus = True, getUrl = True, diskOnly = False,
...
     if active:
-      self.__checkActiveReplicas( result )
+      self.__filterActiveReplicas( result )
     if diskOnly or preferDisk:
       self.__filterTapeReplicas( result, diskOnly = diskOnly )
+    if forJobs:
+      self.__filterReplicasForJobs( result )
     return S_OK( result )
 
+  def getReplicasForJobs( self, lfns, allStatus = False, getUrl = True ):
```
Comment (atsareg) [inline DataManagementSystem/Client/DataManager.py]: Can this function be implemented just in terms of getReplicas( forJobs = True, ... )

Comment (phicharp) [inline DataManagementSystem/Client/DataManager.py]: Yes of course, but Chris' recommendation is to split the functionality off the main function for more clarity... This is a attempt to do so. However I didn't remove the additional parameters from the getReplicas() method.

Comment (chaen) [inline DataManagementSystem/Client/DataManager.py]: I would be in favor of removing the ```forJobs``` from the getReplicas method

File: DataManagementSystem/Client/DataManager.py [hunk 9]
```diff
@@ -1638,13 +1637,27 @@ def getReplicas( self, lfns, allStatus = True, getUrl = True, diskOnly = False,
...
     if active:
-      self.__checkActiveReplicas( result )
+      self.__filterActiveReplicas( result )
     if diskOnly or preferDisk:
       self.__filterTapeReplicas( result, diskOnly = diskOnly )
     if forJobs:
       self.__filterReplicasForJobs( result )
     return S_OK( result )
 
+  def getReplicasForJobs( self, lfns, allStatus = False, getUrl = True ):
```
Comment (fstagni) [inline DataManagementSystem/Client/DataManager.py]: Isn't this equivalent of doing 
```
getReplicas(lfns, allStatus = allStatus, getUrl = getUrl, active = True, preferDisk = True, forJobs = True)
```
?

File: StorageManagementSystem/scripts/dirac-stager-monitor-requests.py [hunk 10]
```diff
@@ -22,160 +22,155 @@
...
+            for tid in tasks:
+              jobs.append( tasks[tid]['SourceTaskID'] )
+            outStr += ' %s ' % ( str( jobs ).ljust( 10 ) )
         else:
-          outStr = '%s %s ' % (outStr, " --- ".ljust(10))     
+          outStr += ' %s ' % ( " --- ".ljust( 10 ) )
       # Stage request info
       # what if there's no request to the site yet?
-      resStageRequests = client.getStageRequests({'ReplicaID':crid})
+      resStageRequests = client.getStageRequests( {'ReplicaID':crid} )
```
Comment (fstagni) [inline StorageManagementSystem/scripts/dirac-stager-monitor-requests.py]: is `crid` defined?

File: StorageManagementSystem/scripts/dirac-stager-monitor-requests.py [hunk 11]
```diff
@@ -22,160 +22,155 @@
...
+      outStr += " %s" % ( str( info['LastUpdate'] ).ljust( 20 ) )
+      outStr += " %s" % ( info['LFN'].ljust( 30 ) )
+      outStr += " %s" % ( info['SE'].ljust( 15 ) )
+      outStr += " %s" % ( str( info['Reason'] ).ljust( 10 ) )
+
       # Task info
-      if 'showJobs' in dictKeys:
-        resTasks = client.getTasks({'ReplicaID':crid})
+      if 'showJobs' in switchDict:
+        resTasks = client.getTasks( {'ReplicaID':crid} )
```
Comment (fstagni) [inline StorageManagementSystem/scripts/dirac-stager-monitor-requests.py]: Also here, is `crid` now `info`?

**Comment by phicharp** [comment]
Yes it would be equivalent, but I got convinced by Chris to not implement it in getReplicas() but have a separate method getReplicasForJobs() that b.t.w. we should start using wherever it is desired...
The man argument is that forJobs=True should imply other flags to be True as well. It is not good to overwrite some flags behind the back of user (e.g. `active = False, forJobs = True` wouldn't make much sense...
I will do this change in another PR...

File: StorageManagementSystem/scripts/dirac-stager-monitor-requests.py (hunk 11 above)
Comment (phicharp) [inline StorageManagementSystem/scripts/dirac-stager-monitor-requests.py]: no, correct, fixing it now

**Comment by fstagni** [comment]
Fine, but you can just, from inside this method, call the other with the "full" signature, instead of repeating the code.

**Comment by fstagni** [comment]
BTW, there are `dirac-stager-monitor-request.py` and `dirac-stager-monitor-requests.py` script, do we need both?

**Comment by phicharp** [comment]
@fstagni: I have no idea and I used neither of those ever ;-) but there were there and I reviewed them

**Comment by phicharp** [comment]
The comment from Chris was that `forJobs` had nothing to do in getReplicas and should be exposed in another method... He managed to convince me and it doesn't harm indeed... 
So now we have 
`* getReplicas()
* getActiveReplicas()
* getReplicasForJobs()`

from the `dirac-dms-lfn-replicas` script point of view, we have something like:

[localhost] ~ $ dirac-dms-lfn-replicas /lhcb/data/2016/RAW/NOBIAS/LHCb/IONPROTON16/187325/187325_0000002099.raw
Successful :
    /lhcb/data/2016/RAW/NOBIAS/LHCb/IONPROTON16/187325/187325_0000002099.raw :
          CERN-RAW : srm://srm-lhcb.cern.ch:8443/srm/managerv2?SFN=/castor/cern.ch/grid/lhcb/data/2016/RAW/NOBIAS/LHCb/IONPROTON16/187325/187325_0000002099.raw
        RAL-BUFFER : srm://srm-lhcb.gridpp.rl.ac.uk:8443/srm/managerv2?SFN=/castor/ads.rl.ac.uk/prod/lhcb/buffer/lhcb/data/2016/RAW/NOBIAS/LHCb/IONPROTON16/187325/187325_0000002099.raw
           RAL-RAW : srm://srm-lhcb.gridpp.rl.ac.uk:8443/srm/managerv2?SFN=/castor/ads.rl.ac.uk/prod/lhcb/data/2016/RAW/NOBIAS/LHCb/IONPROTON16/187325/187325_0000002099.raw
[localhost] ~ $ dirac-dms-lfn-replicas /lhcb/data/2016/RAW/NOBIAS/LHCb/IONPROTON16/187325/187325_0000002099.raw --ForJobs
Successful :
    /lhcb/data/2016/RAW/NOBIAS/LHCb/IONPROTON16/187325/187325_0000002099.raw :
        RAL-BUFFER : srm://srm-lhcb.gridpp.rl.ac.uk:8443/srm/managerv2?SFN=/castor/ads.rl.ac.uk/prod/lhcb/buffer/lhcb/data/2016/RAW/NOBIAS/LHCb/IONPROTON16/187325/187325_0000002099.raw

and for another file that is not on disk:

[localhost] ~ $ dirac-dms-lfn-replicas /lhcb/data/2011/RAW/FULL/LHCb/COLLISION11/93267/093267_0000000058.raw --All
Successful :
    /lhcb/data/2011/RAW/FULL/LHCb/COLLISION11/93267/093267_0000000058.raw :
        CERN-RAW : (AprioriGood) srm://srm-lhcb.cern.ch:8443/srm/managerv2?SFN=/castor/cern.ch/grid/lhcb/data/2011/RAW/FULL/LHCb/COLLISION11/93267/093267_0000000058.raw
         PIC-RAW : (AprioriGood) srm://srmlhcb.pic.es:8443/srm/managerv2?SFN=/pnfs/pic.es/data/lhcb/data/2011/RAW/FULL/LHCb/COLLISION11/93267/093267_0000000058.raw
[localhost] ~ $ dirac-dms-lfn-replicas /lhcb/data/2011/RAW/FULL/LHCb/COLLISION11/93267/093267_0000000058.raw --ForJobs
Successful :
    /lhcb/data/2011/RAW/FULL/LHCb/COLLISION11/93267/093267_0000000058.raw :
        CERN-RAW : srm://srm-lhcb.cern.ch:8443/srm/managerv2?SFN=/castor/cern.ch/grid/lhcb/data/2011/RAW/FULL/LHCb/COLLISION11/93267/093267_0000000058.raw
         PIC-RAW : srm://srmlhcb.pic.es:8443/srm/managerv2?SFN=/pnfs/pic.es/data/lhcb/data/2011/RAW/FULL/LHCb/COLLISION11/93267/093267_0000000058.raw
[localhost] ~ $  dirac-dms-lfn-replicas /lhcb/data/2011/RAW/FULL/LHCb/COLLISION11/93267/093267_0000000058.raw --DiskOnly
Failed :
    /lhcb/data/2011/RAW/FULL/LHCb/COLLISION11/93267/093267_0000000058.raw : No disk replicas

**Comment by phicharp** [comment]
I will make the inventory of the usage of `dm.getReplicas()` and `getActiveReplicas()` and sort them depending on whether we should use `getReplicasForJobs()`, for example in the TS and in the optimizers...

**Pull request merged**